### PR TITLE
Adds Open vSwitch bond management module

### PR DESCRIPTION
**Pull request by busterswt**
##### SUMMARY
Adds a new Open vSwitch bond management module with support for multiple other_config and set commands necessary for multiple interfaces, including DPDK-enabled interfaces.

Based on the existing openvswitch_port module.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

openvswitch_bond

##### ADDITIONAL INFORMATION

The existing Open vSwitch modules do not address the configuration and management of bonded interfaces. This module allows users to create and add bonded interfaces to existing bridges by leveraging the `ovs-vsctl add-bond` command and associated configuration options.

### CODE REVIEW

File: lib/ansible/modules/network/ovs/openvswitch_bond.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+
+DOCUMENTATION = '''
+---
+module: openvswitch_bond
+version_added: 2.8
```
Comment (gundalow) [inline lib/ansible/modules/network/ovs/openvswitch_bond.py]:
```suggestion
version_added: '2.8'
```

File: lib/ansible/modules/network/ovs/openvswitch_bond.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+author: "James Denton (@busterswt)"
+short_description: Manage Open vSwitch bonds
+requirements: [ ovs-vsctl ]
+description:
+    - Manage Open vSwitch bonds and associated options.
+options:
```
Comment (gundalow) [inline lib/ansible/modules/network/ovs/openvswitch_bond.py:29]: In Ansible 2.8 we are trying to improve the module docs by including `type:` could you please add that to all the options

File: lib/ansible/modules/network/ovs/openvswitch_bond.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+    bridge:
+        required: true
```
Comment (gundalow) [inline lib/ansible/modules/network/ovs/openvswitch_bond.py:31]:
```suggestion
        required: true
        type: str
```

File: lib/ansible/modules/network/ovs/openvswitch_bond.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+        description:
+            - Dictionary of other_config applied to a port.
+    set:
+        type: list
+        description:
+            - Sets one or more properties on a port.
+'''
+
+EXAMPLES = '''
+# Creates an active-backup bond using eth4 and eth5 on bridge br-ex
```
Comment (gundalow) [inline lib/ansible/modules/network/ovs/openvswitch_bond.py]: We suggest that people always use a `- name:` when writing Playbooks, doing the same in the EXAMPLES makes it easier for people to copy&paste:
```suggestion
- name: Creates an active-backup bond using eth4 and eth5 on bridge br-ex
```

File: lib/ansible/modules/network/ovs/openvswitch_bond.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+- openvswitch_bond:
```
Comment (gundalow) [inline lib/ansible/modules/network/ovs/openvswitch_bond.py]:
```suggestion
  openvswitch_bond:
```

File: lib/ansible/modules/network/ovs/openvswitch_bond.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+              to prevent flapping.
+    state:
+        default: "present"
+        choices: [ present, absent ]
+        description:
+            - Whether the port should exist
+    timeout:
+        default: 5
+        description:
+            - How long to wait for ovs-vswitchd to respond
```
Comment (gundalow) [inline lib/ansible/modules/network/ovs/openvswitch_bond.py]: Is this seconds?
```suggestion
            - How long to wait for ovs-vswitchd to respond in seconds.
```

**Review by gundalow — COMMENTED** [review]
Hi,
I see from GitHub this is your first PR, congratulations and welcome to the Ansible Community!

I don't know anything about ovs, though I've reviewed the docs and got a few very minor suggestions.

File: lib/ansible/modules/network/ovs/openvswitch_bond.py (hunk 2 above)
Comment (busterswt) [reply lib/ansible/modules/network/ovs/openvswitch_bond.py:29]: Sure can, thanks for the suggestions.

**Review by odyssey4me — APPROVED** [review]
Looks good to me, although I've not tested it. In an ideal world I'd like to see this have a functional test of all its features and functions, perhaps using all the examples. It'd also be double-plus good to see check mode supported so that users can see what will be changed without enacting the changes.

**Comment by kbreit** [comment]
Would you be able to paste an output with `debug` statements so we can see the data structures from the integration tests? I don't have OVS so I can't test it. Overall it looks good but I'd like to see the output.

File: lib/ansible/modules/network/ovs/openvswitch_bond.py [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+        'set': module.params['set']
+    }
+
+    return obj
+
+
+def main():
+    """ Entry point. """
+    argument_spec = {
+        'bridge': {'required': True},
```
Comment (kbreit) [inline lib/ansible/modules/network/ovs/openvswitch_bond.py:334]: My preference is to be explicit about the types for parameters.

File: lib/ansible/modules/network/ovs/openvswitch_bond.py [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+        'bond_downdelay': {'default': None, 'type': 'int'},
+        'state': {'default': 'present', 'choices': ['present', 'absent']},
+        'timeout': {'default': 5, 'type': 'int'},
+        'external_ids': {'default': None, 'type': 'dict'},
+        'other_config': {'default': None, 'type': 'dict'},
+        'set': {'required': False, 'type': 'list', 'default': None}
+    }
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=True)
```
Comment (kbreit) [inline lib/ansible/modules/network/ovs/openvswitch_bond.py:350]: Check mode is preferred (I'm implementing it in my modules now, should have from day 1). If you don't support it, change this to `False`.

**Comment by pabelanger** [comment]
recheck

**Comment by danielmellado** [comment]
Thank you very much for your interest in Ansible. This plugin is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/openvswitch.openvswitch
Please re-submit this PR in the above repository.
If you have further questions please stop by IRC or the mailing list:

IRC: #ansible-network on irc.freenode.net
mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Pull request closed without merging**